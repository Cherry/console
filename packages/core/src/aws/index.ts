import { z } from "zod";
import { zod } from "../util/zod";

export { Account } from "./account";
import { AssumeRoleCommand, STSClient } from "@aws-sdk/client-sts";
import { useWorkspace } from "../actor";
import { useTransaction } from "../util/transaction";
import { awsAccount } from "./aws.sql";
import { and, eq, sql } from "drizzle-orm";

export * as AWS from ".";

const sts = new STSClient({});

export const assumeRole = zod(z.string(), async (id) => {
  const workspaceID = useWorkspace();
  try {
    const result = await sts.send(
      new AssumeRoleCommand({
        RoleArn: `arn:aws:iam::${id}:role/sst-${workspaceID}`,
        RoleSessionName: "sst",
        ExternalId: workspaceID,
        DurationSeconds: 900,
      })
    );
    await useTransaction((tx) =>
      tx
        .update(awsAccount)
        .set({
          timeFailed: null,
        })
        .where(
          and(
            eq(awsAccount.accountID, id),
            eq(awsAccount.workspaceID, workspaceID)
          )
        )
        .execute()
    );
    return {
      secretAccessKey: result.Credentials!.SecretAccessKey!,
      accessKeyId: result.Credentials!.AccessKeyId!,
      sessionToken: result.Credentials!.SessionToken!,
    };
  } catch (e: any) {
    console.error(e);
    await useTransaction((tx) =>
      tx
        .update(awsAccount)
        .set({
          timeFailed: sql`now()`,
        })
        .where(
          and(
            eq(awsAccount.accountID, id),
            eq(awsAccount.workspaceID, workspaceID)
          )
        )
        .execute()
    );
    console.log("failed to assume role for account", id);
    return;
  }
});

export type Credentials = Exclude<
  Awaited<ReturnType<typeof assumeRole>>,
  undefined
>;
