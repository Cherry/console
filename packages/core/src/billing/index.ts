import { createSelectSchema } from "drizzle-zod";
import { usage } from "./billing.sql";
import { z } from "zod";
import { zod } from "../util/zod";
import { createId } from "@paralleldrive/cuid2";
import { eq, and, between, sql } from "drizzle-orm";
import { useTransaction } from "../util/transaction";
import { useWorkspace } from "../actor";
import { workspace } from "../workspace/workspace.sql";
import { Stripe } from "./stripe";
import { DateTime } from "luxon";
import { Warning } from "../warning";

export * as Billing from "./index";
export { Stripe } from "./stripe";

export const Usage = createSelectSchema(usage, {
  id: (schema) => schema.id.cuid2(),
  workspaceID: (schema) => schema.workspaceID.cuid2(),
  stageID: (schema) => schema.stageID.cuid2(),
});
export type Usage = z.infer<typeof Usage>;

const FREE_INVOCATIONS = 1000000;

export const createUsage = zod(
  Usage.pick({ stageID: true, day: true, invocations: true }),
  (input) =>
    useTransaction((tx) =>
      tx
        .insert(usage)
        .values({
          id: createId(),
          workspaceID: useWorkspace(),
          stageID: input.stageID,
          day: input.day,
          invocations: input.invocations,
        })
        .onDuplicateKeyUpdate({
          set: {
            invocations: input.invocations,
          },
        })
        .execute()
    )
);

export const countByStartAndEndDay = zod(
  z.object({
    startDay: Usage.shape.day,
    endDay: Usage.shape.day,
  }),
  async (input) => {
    const rows = await useTransaction((tx) =>
      tx
        .select()
        .from(usage)
        .where(
          and(
            eq(usage.workspaceID, useWorkspace()),
            between(usage.day, input.startDay, input.endDay)
          )
        )
        .execute()
    );
    return rows.reduce((acc, usage) => acc + usage.invocations, 0);
  }
);

export const updateGatingStatus = zod(z.void(), async () => {
  async function isGated() {
    // check subscription status
    const customer = await Stripe.get();
    const subscriptionStatus = customer?.standing;
    if (subscriptionStatus === "overdue") return true;

    const warnings = await Warning.forType({
      type: "permission_usage",
      stageID: "",
    });
    if (warnings.length) return true;

    // check usage
    if (!customer?.subscriptionID) {
      const startDate = DateTime.now().toUTC().startOf("day");
      const invocations = await countByStartAndEndDay({
        startDay: startDate.startOf("month").toSQLDate()!,
        endDay: startDate.endOf("month").toSQLDate()!,
      });
      if (invocations > FREE_INVOCATIONS) return true;
    }
    return false;
  }

  const timeGated = (await isGated()) ? sql`NOW()` : null;

  return useTransaction((tx) =>
    tx
      .update(workspace)
      .set({ timeGated })
      .where(eq(usage.workspaceID, useWorkspace()))
      .execute()
  );
});
