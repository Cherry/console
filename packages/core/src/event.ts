import { createEventBuilder } from "sst/node/event-bus";
import { useActor } from "./actor";

export const event = createEventBuilder({
  bus: "bus",
  metadataFn() {
    return {
      actor: useActor(),
    };
  },
});
