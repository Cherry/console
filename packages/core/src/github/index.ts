import { z } from "zod";
import { zod } from "../util/zod";
import { App } from "octokit";
import { useTransaction } from "../util/transaction";
import { githubOrg } from "./github.sql";
import { useWorkspace } from "../actor";
import { createId } from "@paralleldrive/cuid2";
import { and, db, eq } from "../drizzle";
import { createSelectSchema } from "drizzle-zod";
import { Config } from "sst/node/config";

export * as Github from "./index";

export const Info = createSelectSchema(githubOrg);
export type Info = z.infer<typeof Info>;

const app = new App({
  appId: Config.GITHUB_APP_ID,
  privateKey: Config.GITHUB_PRIVATE_KEY,
});

export const connect = zod(z.number(), async (installationID) => {
  // Get installation detail
  const octokit = await app.getInstallationOctokit(installationID);
  const installation = await octokit.rest.apps.getInstallation({
    installation_id: installationID,
  });
  const orgID = installation.data.account?.id!;
  const orgSlug =
    "slug" in installation.data.account!
      ? installation.data.account!.slug
      : installation.data.account!.login;

  await useTransaction(async (tx) =>
    tx
      .insert(githubOrg)
      .values({
        workspaceID: useWorkspace(),
        id: createId(),
        orgID,
        orgSlug,
        installationID,
      })
      .onDuplicateKeyUpdate({
        set: {
          installationID,
          orgSlug,
        },
      })
      .execute()
  );
});

export const disconnect = zod(Info.shape.id, (input) =>
  useTransaction((tx) => {
    return tx
      .delete(githubOrg)
      .where(
        and(eq(githubOrg.id, input), eq(githubOrg.workspaceID, useWorkspace()))
      )
      .execute();
  })
);

export const disconnectAll = zod(z.number().int(), (input) =>
  useTransaction((tx) => {
    return tx.delete(githubOrg).where(eq(githubOrg.orgID, input)).execute();
  })
);

// TODO remove
//export const send = zod(
//  z.object({
//    channel: z.string().nonempty(),
//    text: z.string().nonempty(),
//    blocks: z.custom<KnownBlock[]>(),
//  }),
//  async (input) => {
//    const result = await db
//      .select()
//      .from(githubOrg)
//      .where(and(eq(githubOrg.workspaceID, useWorkspace())))
//      .limit(1)
//      .then((rows) => rows.at(0));
//
//    if (!result) return;
//
//    const client = new WebClient(result.accessToken);
//    await client.chat.postMessage({
//      blocks: input.blocks,
//      unfurl_links: false,
//      text: input.text,
//      channel: input.channel,
//    });
//  }
//);
