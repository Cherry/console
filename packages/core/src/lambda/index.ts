import { z } from "zod";
import { zod } from "../util/zod";
import { InvokeCommand, LambdaClient } from "@aws-sdk/client-lambda";
import { Stage } from "../app";
import { useTransaction } from "../util/transaction";
import { lambdaPayload } from "./lambda.sql";
import { useActor, useWorkspace } from "../actor";
import { createSelectSchema } from "drizzle-zod";
import { createId } from "@paralleldrive/cuid2";

export * as Lambda from ".";

export const invoke = zod(
  z.object({
    stageID: z.string(),
    functionARN: z.string(),
    payload: z.any(),
  }),
  async (input) => {
    const config = await Stage.assumeRole(input.stageID);
    const client = new LambdaClient(config);
    await client.send(
      new InvokeCommand({
        FunctionName: input.functionARN,
        Payload: Buffer.from(JSON.stringify(input.payload)),
        InvocationType: "Event",
      })
    );
  }
);

export const SavedPayload = createSelectSchema(lambdaPayload, {
  id: (schema) => schema.id.cuid2(),
});

export const savePayload = zod(
  SavedPayload.pick({
    id: true,
    name: true,
    functionARN: true,
    payload: true,
  }).partial({
    id: true,
  }),
  async (input) =>
    useTransaction(async (tx) => {
      const id = input.id || createId();
      await tx.insert(lambdaPayload).values({
        id,
        functionARN: input.functionARN,
        name: input.name,
        creator: useActor(),
        workspaceID: useWorkspace(),
        payload: input.payload,
      });
    })
);
