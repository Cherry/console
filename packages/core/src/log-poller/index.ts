import { createSelectSchema } from "drizzle-zod";
import { log_poller } from "./log-poller.sql";
import { zod } from "../util/zod";
import { useTransaction } from "../util/transaction";
import cuid2 from "@paralleldrive/cuid2";
import { eq } from "drizzle-orm";

export const Info = createSelectSchema(log_poller, {
  id: (schema) => schema.id.cuid2(),
});

export const subscribe = zod(Info.shape.logGroup, async (logGroup) =>
  useTransaction(async (tx) => {
    const result = await tx
      .insert(log_poller)
      .values({
        logGroup,
        id: cuid2.createId(),
      })
      .onDuplicateKeyUpdate({
        set: {},
      })
      .execute();
    console.log("subscribe result", result);
    const id = await tx
      .select({
        id: log_poller.id,
      })
      .from(log_poller)
      .where(eq(log_poller.logGroup, logGroup))
      .execute()
      .then((rows) => rows[0]!.id);
  })
);
