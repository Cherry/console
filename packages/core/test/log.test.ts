import { expect, test } from "vitest";
import { Log } from "../src/log";

function expectError(input: string[]) {
  const err = Log.extractError(input);
  expect(err).not.toBeUndefined();
  expect(err).toMatchSnapshot();
}

test("node invoke", () => {
  expectError([
    `2023-09-15T13:03:59.257Z`,
    `8179a9db-e726-451b-898f-93923b3e7bbd`,
    `ERROR`,
    `Invoke Error`,
    `{"errorType":"NoSuchBucket","errorMessage":"The specified bucket does not exist","name":"NoSuchBucket","$fault":"client","$metadata":{"httpStatusCode":404,"requestId":"SEDJ1KBFNG26CADF","extendedRequestId":"Uxm/2OOqe5zjcRwxzZNzpePcai5oteEZJoq1S7gTWF94BO1ClmQAl5nYM7UeOpMZKAla5SFqJdI=","attempts":1,"totalRetryDelay":0},"Code":"NoSuchBucket","BucketName":"sstbootstrap-useast124d14e4b-txrftoxsysmw","RequestId":"SEDJ1KBFNG26CADF","HostId":"Uxm/2OOqe5zjcRwxzZNzpePcai5oteEZJoq1S7gTWF94BO1ClmQAl5nYM7UeOpMZKAla5SFqJdI=","message":"The specified bucket does not exist","stack":["NoSuchBucket: The specified bucket does not exist"," at de_NoSuchBucketRes (file:///var/task/packages/functions/src/events/app-stage-updated.mjs:51859:21)"," at de_ListObjectsV2CommandError (file:///var/task/packages/functions/src/events/app-stage-updated.mjs:51807:19)"," at runMicrotasks (<anonymous>)"," at processTicksAndRejections (node:internal/process/task_queues:96:5)"," at async file:///var/task/packages/functions/src/events/app-stage-updated.mjs:42732:20"," at async file:///var/task/packages/functions/src/events/app-stage-updated.mjs:41547:18"," at async file:///var/task/packages/functions/src/events/app-stage-updated.mjs:43230:38"," at async file:///var/task/packages/functions/src/events/app-stage-updated.mjs:41269:22"," at async file:///var/task/packages/functions/src/events/app-stage-updated.mjs:56745:19"," at async file:///var/task/packages/functions/src/events/app-stage-updated.mjs:56891:3"]}`,
  ]);
});

test("node uncaught exception", () => {
  expectError([
    `2023-09-15T13:03:59.257Z`,
    `8179a9db-e726-451b-898f-93923b3e7bbd`,
    `ERROR`,
    `Uncaught Exception`,
    `{"errorType":"NoSuchBucket","errorMessage":"The specified bucket does not exist","name":"NoSuchBucket","$fault":"client","$metadata":{"httpStatusCode":404,"requestId":"SEDJ1KBFNG26CADF","extendedRequestId":"Uxm/2OOqe5zjcRwxzZNzpePcai5oteEZJoq1S7gTWF94BO1ClmQAl5nYM7UeOpMZKAla5SFqJdI=","attempts":1,"totalRetryDelay":0},"Code":"NoSuchBucket","BucketName":"sstbootstrap-useast124d14e4b-txrftoxsysmw","RequestId":"SEDJ1KBFNG26CADF","HostId":"Uxm/2OOqe5zjcRwxzZNzpePcai5oteEZJoq1S7gTWF94BO1ClmQAl5nYM7UeOpMZKAla5SFqJdI=","message":"The specified bucket does not exist","stack":["NoSuchBucket: The specified bucket does not exist"," at de_NoSuchBucketRes (file:///var/task/packages/functions/src/events/app-stage-updated.mjs:51859:21)"," at de_ListObjectsV2CommandError (file:///var/task/packages/functions/src/events/app-stage-updated.mjs:51807:19)"," at runMicrotasks (<anonymous>)"," at processTicksAndRejections (node:internal/process/task_queues:96:5)"," at async file:///var/task/packages/functions/src/events/app-stage-updated.mjs:42732:20"," at async file:///var/task/packages/functions/src/events/app-stage-updated.mjs:41547:18"," at async file:///var/task/packages/functions/src/events/app-stage-updated.mjs:43230:38"," at async file:///var/task/packages/functions/src/events/app-stage-updated.mjs:41269:22"," at async file:///var/task/packages/functions/src/events/app-stage-updated.mjs:56745:19"," at async file:///var/task/packages/functions/src/events/app-stage-updated.mjs:56891:3"]}`,
  ]);
});

test("node inline multiline", () => {
  expectError([
    `2023-09-15T12:43:23.014Z`,
    `31eaad44-5f5e-41ae-ae77-db15c7f7c8a2`,
    `ERROR`,
    `ZodError: [\n  {\n    "code": "too_small",\n    "minimum": 1,\n    "type": "string",\n    "inclusive": true,\n    "exact": false,\n    "message": "String must contain at least 1 character(s)",\n    "path": [\n      "slug"\n    ]\n  },\n  {\n    "code": "too_small",\n    "minimum": 3,\n    "type": "string",\n    "inclusive": true,\n    "exact": false,\n    "message": "String must contain at least 3 character(s)",\n    "path": [\n      "slug"\n    ]\n  },\n  {\n    "validation": "regex",\n    "code": "invalid_string",\n    "message": "Invalid",\n    "path": [\n      "slug"\n    ]\n  }\n]\n    at Object.get error [as error] (file:///var/task/packages/functions/src/replicache/push1.mjs:37260:23)\n    at _ZodObject.parse (file:///var/task/packages/functions/src/replicache/push1.mjs:37363:18)\n    at Object.result (file:///var/task/packages/functions/src/replicache/push1.mjs:58013:27)\n    at file:///var/task/packages/functions/src/replicache/push1.mjs:76714:46\n    at Object.fn (file:///var/task/packages/functions/src/replicache/push1.mjs:58262:16)\n    at Server.execute (file:///var/task/packages/functions/src/replicache/push1.mjs:58279:16)\n    at file:///var/task/packages/functions/src/replicache/push1.mjs:76784:22\n    at runMicrotasks (<anonymous>)\n    at processTicksAndRejections (node:internal/process/task_queues:96:5)\n    at async db.transaction.isolationLevel (file:///var/task/packages/functions/src/replicache/push1.mjs:55329:25) {\n  issues: [\n    {\n      code: \'too_small\',\n      minimum: 1,\n      type: \'string\',\n      inclusive: true,\n      exact: false,\n      message: \'String must contain at least 1 character(s)\',\n      path: [Array]\n    },\n    {\n      code: \'too_small\',\n      minimum: 3,\n      type: \'string\',\n      inclusive: true,\n      exact: false,\n      message: \'String must contain at least 3 character(s)\',\n      path: [Array]\n    },\n    {\n      validation: \'regex\',\n      code: \'invalid_string\',\n      message: \'Invalid\',\n      path: [Array]\n    }\n  ],\n  addIssue: [Function (anonymous)],\n  addIssues: [Function (anonymous)],\n  errors: [\n    {\n      code: \'too_small\',\n      minimum: 1,\n      type: \'string\',\n      inclusive: true,\n      exact: false,\n      message: \'String must contain at least 1 character(s)\',\n      path: [Array]\n    },\n    {\n      code: \'too_small\',\n      minimum: 3,\n      type: \'string\',\n      inclusive: true,\n      exact: false,\n      message: \'String must contain at least 3 character(s)\',\n      path: [Array]\n    },\n    {\n      validation: \'regex\',\n      code: \'invalid_string\',\n      message: \'Invalid\',\n      path: [Array]\n    }\n  ]\n}\n`,
  ]);
});

test("node inline", () => {
  expectError([
    `2023-09-12T00:55:20.974Z`,
    `662fa6b4-dc47-4d49-a7e2-200ad7ce1537`,
    `ERROR`,
    `DatabaseError: target: sst.-.primary: vttablet: rpc error: code = AlreadyExists desc = Duplicate entry 'foo' for key 'workspace.slug' (errno 1062) (sqlstate 23000) (CallerID: lq15hm6s0ctusshuarsx): Sql: \"insert into workspace(id, time_created, time_updated, time_deleted, slug, stripe_customer_id, stripe_subscription_id, stripe_subscription_item_id) values (:vtg1 /* VARCHAR */, default, default, default, :vtg2 /* VARCHAR */, default, default, default)\", BindVars: {REDACTED}\n    at _Connection.execute (file:///var/task/packages/functions/src/replicache/push1.mjs:47211:13)\n    at runMicrotasks (<anonymous>)\n    at processTicksAndRejections (node:internal/process/task_queues:96:5)\n    at async file:///var/task/packages/functions/src/replicache/push1.mjs:69523:5\n    at async file:///var/task/packages/functions/src/replicache/push1.mjs:76353:22\n    at async file:///var/task/packages/functions/src/replicache/push1.mjs:76423:9\n    at async db.transaction.isolationLevel (file:///var/task/packages/functions/src/replicache/push1.mjs:55325:25)\n    at async _Connection.transaction (file:///var/task/packages/functions/src/replicache/push1.mjs:47190:19)\n    at async createTransaction (file:///var/task/packages/functions/src/replicache/push1.mjs:55322:20)\n    at async file:///var/task/packages/functions/src/replicache/push1.mjs:76380:5 {\n  status: 400,\n  body: {\n    message: \`target: sst.-.primary: vttablet: rpc error: code = AlreadyExists desc = Duplicate entry 'foo' for key 'workspace.slug' (errno 1062) (sqlstate 23000) (CallerID: lq15hm6s0ctusshuarsx): Sql: \"insert into workspace(id, time_created, time_updated, time_deleted, slug, stripe_customer_id, stripe_subscription_id, stripe_subscription_item_id) values (:vtg1 /* VARCHAR */, default, default, default, :vtg2 /* VARCHAR */, default, default, default)\", BindVars: {REDACTED}\`,\n    code: 'UNKNOWN'\n  }\n}\n`,
  ]);
});

test("node ignore warning", () => {
  const err = Log.extractError([
    `2023-09-12T00:55:20.974Z`,
    `662fa6b4-dc47-4d49-a7e2-200ad7ce1537`,
    `ERROR`,
    `(node:8) [DEP0005] DeprecationWarning: Buffer() is deprecated due to security and usability issues. Please use the Buffer.alloc(), Buffer.allocUnsafe(), or Buffer.from() methods instead.
(Use \`node --trace-deprecation ...\` to show where the warning was created)`,
  ]);
  expect(err).toBeUndefined();
});

test("node logtail", () => {
  expectError([
    `2023-09-12T00:55:20.974Z`,
    `662fa6b4-dc47-4d49-a7e2-200ad7ce1537`,
    `ERROR`,
    "{\n" +
      "  status: 500,\n" +
      `  message: "Cannot read properties of undefined (reading 'enabled')",\n` +
      "  stack: [\n" +
      `    "TypeError: Cannot read properties of undefined (reading 'enabled')",\n` +
      "    'at Object.toJSON (file:///var/task/src/handlers/http.mjs:280853:35)',\n" +
      "    'at transform (file:///var/task/src/handlers/http.mjs:280495:30)',\n" +
      "    'at Document2.$toObject (file:///var/task/src/handlers/http.mjs:91609:25)',\n" +
      "    'at Document2.toJSON (file:///var/task/src/handlers/http.mjs:91810:19)',\n" +
      "    'at clone (file:///var/task/src/handlers/http.mjs:83888:21)',\n" +
      "    'at cloneObject (file:///var/task/src/handlers/http.mjs:83960:22)',\n" +
      "    'at clone (file:///var/task/src/handlers/http.mjs:83901:20)',\n" +
      "    'at Document2.$toObject (file:///var/task/src/handlers/http.mjs:91580:17)',\n" +
      "    'at Document2.toJSON (file:///var/task/src/handlers/http.mjs:91810:19)',\n" +
      "    'at RESPONSE.stringify [as _serializer] (<anonymous>)'\n" +
      "  ]\n" +
      "} {\n" +
      "  context: {\n" +
      "    runtime: {\n" +
      "      file: '/var/task/src/handlers/http.mjs',\n" +
      "      type: 'Object',\n" +
      "      method: 'error',\n" +
      "      function: 'error',\n" +
      "      line: 280278,\n" +
      "      column: 15\n" +
      "    },\n" +
      "    system: { pid: 8, main_file: '' }\n" +
      "  },\n" +
      "  service: 'http',\n" +
      "  id: '2f78a09a'\n" +
      "}",
  ]);
});

test("node logtail", () => {
  expectError([
    `2023-09-12T00:55:20.974Z`,
    `662fa6b4-dc47-4d49-a7e2-200ad7ce1537`,
    `ERROR`,
    "{\n" +
      "  status: 500,\n" +
      `  message: "Cannot read properties of undefined (reading 'enabled')",\n` +
      "  stack: [\n" +
      `    "TypeError: Cannot read properties of undefined (reading 'enabled')",\n` +
      "    'at Object.toJSON (file:///var/task/src/handlers/http.mjs:280853:35)',\n" +
      "    'at transform (file:///var/task/src/handlers/http.mjs:280495:30)',\n" +
      "    'at Document2.$toObject (file:///var/task/src/handlers/http.mjs:91609:25)',\n" +
      "    'at Document2.toJSON (file:///var/task/src/handlers/http.mjs:91810:19)',\n" +
      "    'at clone (file:///var/task/src/handlers/http.mjs:83888:21)',\n" +
      "    'at cloneObject (file:///var/task/src/handlers/http.mjs:83960:22)',\n" +
      "    'at clone (file:///var/task/src/handlers/http.mjs:83901:20)',\n" +
      "    'at Document2.$toObject (file:///var/task/src/handlers/http.mjs:91580:17)',\n" +
      "    'at Document2.toJSON (file:///var/task/src/handlers/http.mjs:91810:19)',\n" +
      "    'at RESPONSE.stringify [as _serializer] (<anonymous>)'\n" +
      "  ]\n" +
      "} {\n" +
      "  context: {\n" +
      "    runtime: {\n" +
      "      file: '/var/task/src/handlers/http.mjs',\n" +
      "      type: 'Object',\n" +
      "      method: 'error',\n" +
      "      function: 'error',\n" +
      "      line: 280278,\n" +
      "      column: 15\n" +
      "    },\n" +
      "    system: { pid: 8, main_file: '' }\n" +
      "  },\n" +
      "  service: 'http',\n" +
      "  id: '2f78a09a'\n" +
      "}",
  ]);
});

test("node timeout", () => {
  expectError([
    `2023-09-12T00:55:20.974Z`,
    `662fa6b4-dc47-4d49-a7e2-200ad7ce1537`,
    `Task timed out after 120.12 seconds`,
  ]);
});

test("node undefined requestID", () =>
  expectError([
    `2023-09-12T00:55:20.974Z`,
    `undefined`,
    `ERROR`,
    "Error: Missing AWS Lambda trace data for X-Ray. Ensure Active Tracing is enabled and no subsegments are created outside the function handler.\n" +
      "    at Object.contextMissingLogError [as contextMissing] (file:///var/task/functions/crons/property/queue-property-updates.js:738560:23)\n" +
      "    at Segment.resolveLambdaTraceData (file:///var/task/functions/crons/property/queue-property-updates.js:739641:47)\n" +
      "    at Object.getSegment (file:///var/task/functions/crons/property/queue-property-updates.js:738604:21)\n" +
      "    at tryGetCurrentSegment (file:///var/task/functions/crons/property/queue-property-updates.js:742664:29)\n" +
      "    at Promise2.then (file:///var/task/functions/crons/property/queue-property-updates.js:742631:49)\n" +
      "    at file:///var/task/functions/crons/property/queue-property-updates.js:820938:34\n" +
      "    at step (file:///var/task/functions/crons/property/queue-property-updates.js:5739:25)\n" +
      "    at Object.next (file:///var/task/functions/crons/property/queue-property-updates.js:5686:20)\n" +
      "    at file:///var/task/functions/crons/property/queue-property-updates.js:5672:73\n" +
      "    at new Promise (<anonymous>)",
  ]));
