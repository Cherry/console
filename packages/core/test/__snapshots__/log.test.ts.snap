// Vitest Snapshot v1

exports[`node inline 1`] = `
{
  "error": "DatabaseError",
  "message": "target: sst.-.primary: vttablet: rpc error: code = AlreadyExists desc = Duplicate entry 'foo' for key 'workspace.slug' (errno 1062) (sqlstate 23000) (CallerID: lq15hm6s0ctusshuarsx): Sql: \\"insert into workspace(id, time_created, time_updated, time_deleted, slug, stripe_customer_id, stripe_subscription_id, stripe_subscription_item_id) values (:vtg1 /* VARCHAR */, default, default, default, :vtg2 /* VARCHAR */, default, default, default)\\", BindVars: {REDACTED}",
  "stack": [
    {
      "raw": "at _Connection.execute (file:///var/task/packages/functions/src/replicache/push1.mjs:47211:13)",
    },
    {
      "raw": "at runMicrotasks (<anonymous>)",
    },
    {
      "raw": "at processTicksAndRejections (node:internal/process/task_queues:96:5)",
    },
    {
      "raw": "at async file:///var/task/packages/functions/src/replicache/push1.mjs:69523:5",
    },
    {
      "raw": "at async file:///var/task/packages/functions/src/replicache/push1.mjs:76353:22",
    },
    {
      "raw": "at async file:///var/task/packages/functions/src/replicache/push1.mjs:76423:9",
    },
    {
      "raw": "at async db.transaction.isolationLevel (file:///var/task/packages/functions/src/replicache/push1.mjs:55325:25)",
    },
    {
      "raw": "at async _Connection.transaction (file:///var/task/packages/functions/src/replicache/push1.mjs:47190:19)",
    },
    {
      "raw": "at async createTransaction (file:///var/task/packages/functions/src/replicache/push1.mjs:55322:20)",
    },
    {
      "raw": "at async file:///var/task/packages/functions/src/replicache/push1.mjs:76380:5 {
  status: 400,
  body: {
    message: \`target: sst.-.primary: vttablet: rpc error: code = AlreadyExists desc = Duplicate entry 'foo' for key 'workspace.slug' (errno 1062) (sqlstate 23000) (CallerID: lq15hm6s0ctusshuarsx): Sql: \\"insert into workspace(id, time_created, time_updated, time_deleted, slug, stripe_customer_id, stripe_subscription_id, stripe_subscription_item_id) values (:vtg1 /* VARCHAR */, default, default, default, :vtg2 /* VARCHAR */, default, default, default)\\", BindVars: {REDACTED}\`,
    code: 'UNKNOWN'
  }
}",
    },
  ],
}
`;

exports[`node inline multiline 1`] = `
{
  "error": "ZodError",
  "message": "[
  {
    \\"code\\": \\"too_small\\",
    \\"minimum\\": 1,
    \\"type\\": \\"string\\",
    \\"inclusive\\": true,
    \\"exact\\": false,
    \\"message\\": \\"String must contain at least 1 character(s)\\",
    \\"path\\": [
      \\"slug\\"
    ]
  },
  {
    \\"code\\": \\"too_small\\",
    \\"minimum\\": 3,
    \\"type\\": \\"string\\",
    \\"inclusive\\": true,
    \\"exact\\": false,
    \\"message\\": \\"String must contain at least 3 character(s)\\",
    \\"path\\": [
      \\"slug\\"
    ]
  },
  {
    \\"validation\\": \\"regex\\",
    \\"code\\": \\"invalid_string\\",
    \\"message\\": \\"Invalid\\",
    \\"path\\": [
      \\"slug\\"
    ]
  }
]",
  "stack": [
    {
      "raw": "at Object.get error [as error] (file:///var/task/packages/functions/src/replicache/push1.mjs:37260:23)",
    },
    {
      "raw": "at _ZodObject.parse (file:///var/task/packages/functions/src/replicache/push1.mjs:37363:18)",
    },
    {
      "raw": "at Object.result (file:///var/task/packages/functions/src/replicache/push1.mjs:58013:27)",
    },
    {
      "raw": "at file:///var/task/packages/functions/src/replicache/push1.mjs:76714:46",
    },
    {
      "raw": "at Object.fn (file:///var/task/packages/functions/src/replicache/push1.mjs:58262:16)",
    },
    {
      "raw": "at Server.execute (file:///var/task/packages/functions/src/replicache/push1.mjs:58279:16)",
    },
    {
      "raw": "at file:///var/task/packages/functions/src/replicache/push1.mjs:76784:22",
    },
    {
      "raw": "at runMicrotasks (<anonymous>)",
    },
    {
      "raw": "at processTicksAndRejections (node:internal/process/task_queues:96:5)",
    },
    {
      "raw": "at async db.transaction.isolationLevel (file:///var/task/packages/functions/src/replicache/push1.mjs:55329:25) {
  issues: [
    {
      code: 'too_small',
      minimum: 1,
      type: 'string',
      inclusive: true,
      exact: false,
      message: 'String must contain at least 1 character(s)',
      path: [Array]
    },
    {
      code: 'too_small',
      minimum: 3,
      type: 'string',
      inclusive: true,
      exact: false,
      message: 'String must contain at least 3 character(s)',
      path: [Array]
    },
    {
      validation: 'regex',
      code: 'invalid_string',
      message: 'Invalid',
      path: [Array]
    }
  ],
  addIssue: [Function (anonymous)],
  addIssues: [Function (anonymous)],
  errors: [
    {
      code: 'too_small',
      minimum: 1,
      type: 'string',
      inclusive: true,
      exact: false,
      message: 'String must contain at least 1 character(s)',
      path: [Array]
    },
    {
      code: 'too_small',
      minimum: 3,
      type: 'string',
      inclusive: true,
      exact: false,
      message: 'String must contain at least 3 character(s)',
      path: [Array]
    },
    {
      validation: 'regex',
      code: 'invalid_string',
      message: 'Invalid',
      path: [Array]
    }
  ]
}",
    },
  ],
}
`;

exports[`node invoke 1`] = `
{
  "error": "NoSuchBucket",
  "message": "The specified bucket does not exist",
  "stack": [
    {
      "raw": "at de_NoSuchBucketRes (file:///var/task/packages/functions/src/events/app-stage-updated.mjs:51859:21)",
    },
    {
      "raw": "at de_ListObjectsV2CommandError (file:///var/task/packages/functions/src/events/app-stage-updated.mjs:51807:19)",
    },
    {
      "raw": "at runMicrotasks (<anonymous>)",
    },
    {
      "raw": "at processTicksAndRejections (node:internal/process/task_queues:96:5)",
    },
    {
      "raw": "at async file:///var/task/packages/functions/src/events/app-stage-updated.mjs:42732:20",
    },
    {
      "raw": "at async file:///var/task/packages/functions/src/events/app-stage-updated.mjs:41547:18",
    },
    {
      "raw": "at async file:///var/task/packages/functions/src/events/app-stage-updated.mjs:43230:38",
    },
    {
      "raw": "at async file:///var/task/packages/functions/src/events/app-stage-updated.mjs:41269:22",
    },
    {
      "raw": "at async file:///var/task/packages/functions/src/events/app-stage-updated.mjs:56745:19",
    },
    {
      "raw": "at async file:///var/task/packages/functions/src/events/app-stage-updated.mjs:56891:3",
    },
  ],
}
`;

exports[`node uncaught exception 1`] = `
{
  "error": "NoSuchBucket",
  "message": "The specified bucket does not exist",
  "stack": [
    {
      "raw": "at de_NoSuchBucketRes (file:///var/task/packages/functions/src/events/app-stage-updated.mjs:51859:21)",
    },
    {
      "raw": "at de_ListObjectsV2CommandError (file:///var/task/packages/functions/src/events/app-stage-updated.mjs:51807:19)",
    },
    {
      "raw": "at runMicrotasks (<anonymous>)",
    },
    {
      "raw": "at processTicksAndRejections (node:internal/process/task_queues:96:5)",
    },
    {
      "raw": "at async file:///var/task/packages/functions/src/events/app-stage-updated.mjs:42732:20",
    },
    {
      "raw": "at async file:///var/task/packages/functions/src/events/app-stage-updated.mjs:41547:18",
    },
    {
      "raw": "at async file:///var/task/packages/functions/src/events/app-stage-updated.mjs:43230:38",
    },
    {
      "raw": "at async file:///var/task/packages/functions/src/events/app-stage-updated.mjs:41269:22",
    },
    {
      "raw": "at async file:///var/task/packages/functions/src/events/app-stage-updated.mjs:56745:19",
    },
    {
      "raw": "at async file:///var/task/packages/functions/src/events/app-stage-updated.mjs:56891:3",
    },
  ],
}
`;
